### DISCUSSION TITLE
messaging().setBackgroundMessageHandler not trigger on iOs when app is in background state and dev mode only

### DISCUSSION THREAD

**Discussion by ND-creator** [invertase/react-native-firebase#6889] [Q&A]
Hello,

I implemented push notification (data-only/silent notif) with notifee on client side, and fcm firebase-admin on nodejs server side. It works fine on Android, and on iOs it works fine with TestFlight but in dev mode messaging().setBackgroundMessageHandler() do not trigger when the app is in background or quit state, i receive push notif only in foreground state. Do you have any idea how to resolve it please ?

Thank you, have a good day

**Comment by Sergey-lang** [comment]
My push delivered but I can't sent http request to my server because background handler not triggered. On android I can.
 It would be interesting to know how do it.

**Reply by mikehardy** [reply]
background message delivery on iOS is not guaranteed, by design, by Apple in defense of battery life on their platform.
If you are supplying the content-available flag set to 1 in the JSON, and the device has enough power, and the device is not overheated, and the operating system thinks you have used the app enough that the user wants to spend battery on it, and background refresh is enabled for the app, and power saving is not turned on for the device, *then* the message may deliver in the background on iOS.

You can watch the iOS logs to see the decision making process the operating system does as it decides whether to deliver the message or not.

**Reply by Sergey-lang** [reply]
@mikehardy  thanks for the explanation, Mike)

**Reply by ND-creator** [reply]
how does whatsapp do ? they do not use data-only ? i receive whatsapp message even if the battery is low

**Reply by mikehardy** [reply]
You might be interested in "Notification Extensions" in iOS. Related: https://notifee.app/react-native/docs/ios/remote-notification-support (sorry, to answer your question directly: they do not use data-only, *no one that needs reliable delivery uses data-only* - they use notification blocks and have extensions that give them a limited amount of time to process the notification before display to users)

**Reply by Sergey-lang** [reply]
this extension calls only with "notifications" or "alert" block, but it's not clear what to do with it) It's call in objective-c code) If I want to send request about delivery I need to know objective-c? @mikehardy

**Reply by mikehardy** [reply]
If you want to develop mobile applications with notifications that are not basic, then some native platform knowledge is a requirement. Apple is a difficult platform.

**Reply by Sergey-lang** [reply]
I found solution. I made swift extention file instead objective-c. It's more clearly fo me. I add https request(google help) in didRecieved method and it's help me to get feedback about message delivered status. 
@mikehardy

**Comment by Sergey-lang** [comment]
![image](https://user-images.githubusercontent.com/63201050/217552770-6a7e1f11-7ac2-4547-bc3a-79bf96e0079c.png)
code above called yesterday when even app not run(andoid)) today it's not work for android too.
simple request to test api 
` await axios.post('https://webhook.site/1234124', {
      messageId: dto.messageId,
      status: dto.status,
    });`

https://fcm.googleapis.com/v1/projects/test/messages:send

JSON
`{
   "message":{
      "token": "",
      "notification":{
         "title": "test 1",
         "body":"test 2"
      },
      "apns": {
        "headers": {
            "apns-priority": "5"
        },
        "payload": {
            "aps": {
                "contentAvailable": 1,
                "sound": "default"
            }
        }
    },
      "data":{
        "link":"https://test.com"
      }
   }
}`

@mikehardy Could you get some tips for me? I would be very grateful, maybe I missed something. I used old FCM API yesterday
